### DISCUSSION TITLE
Feedback - Swiss e-ID and trust infrastructure: Trust protocol based on VCs

### DISCUSSION THREAD

**Discussion by bj-ms** [e-id-admin/open-source-community#33] [General]
This thread is for feedback, questions or general inputs on the new trust protocol based on VCs specification.

Link to the new specification: https://github.com/e-id-admin/open-source-community/blob/main/tech-roadmap/rfcs/trust-protocol/trust-protocol.md

**Comment by mdoujak** [comment]
I like the idea of creating a trust protocol with VCs. I am just wondering on whether the proposal meets the expectations. To clarify my concerns, I have to provide a (probably incomplete) list of requirements:

1. To identify an issuer, I would like to receive a set of claims that is useful in the real world:
- Name
- Address
- Contact Information (phone, email,...)
- Web Site
- Legal references (e.g. Data Privacy Policy, ...)
2. To identify the authority of the entity that issued the Trust VC for the Issuer, I would expect the authority VC to be embedded in trust VC. This could be done recursively to create intermediate authorities and longer chains.
3. The claims of the authority would be similar or identical to the claims of the issuer.
4. Authority VCs should also carry "authorization" information. E.g. it could define, what types of VCs the authority is allowed to issue (Issuer VC and/or Authority VCs and/or Verifier VCs)
5. Issuer VCs could optionally contain a list of Schema references for VC Schemas this Issuer is able/allowed to issue. This could be combined with policy information (e.g. the Issuer is only allowed to issue a VC Schema of a specific Version, all Version issued after version X, or only the newest version.
6. Verifier VCs would have similar claims to the issuer VCs.
7. Verifier VCs should contain references to the presentation schemas that this verifier is able/allowed to request from the Holder.
8. As a Holder, I could then just have a list of dids of the root authorities of the ecosystems that I want to trust in my wallet and everything else could be derived from these "trust anchors". Each trust anchor would correspond with an ecosystem.


Example: 
1. Santesuisse could decide to create an ecosystem for healthcare insurance in Switzerland. They would control the "Trust Anchor VC of the healthcare insurance" ecosystem.
2. SASIS could be the authorized intermediary to manage the technical part of the VC Issuance for the healthcare insurances.
3. As a Holder it is sufficient to trust Santesuisse to trust all the healthcare insurances.

Example 2:
1. BAG decides to create a healthcare ecosystem and they control the trust anchor of this ecosystem.
2. Santesuisse decides that healthcare insurance should be part of healthcare and they ask for Trust VC as a intermediate that can issue VCs for authorities, issuers and verifiers.
4. SASIS can retain their role to manage the technical parts of VC issuance.
5. HIN decides that their organization is also part of the healthcare ecosystem and they will receive a similar VC like Santesuisse.
6. As a Holder it is now sufficient to trust BAG to have all the doctors and all the healthcare insurances as trustworthy in the wallet.

**Reply by AnAnalogGuy** [reply]
From an application perspective (not from a solution point of view), aren't you describing what a PKI normally would do? Building a chain of trust?

**Reply by mdoujak** [reply]
Yes, of course. But then using hierarchical systems is not unique to PKI.

The main difference is that the indirection through the TSP is not required. With the governance body being the issuer, the entire registration part has likely been done, and the VC issuance is just a minor thing.

Example: To operate in the Swiss financial market, FINMA must accredit you. If FINMA wanted to issue VCs to banks, this would be relatively simple since the register of all the banks or FinTechs already exists.

**Reply by AnAnalogGuy** [reply]
You point out exactly the problem: as far as I understand, the governance body is only doing a registration in terms of an onboarding into the system, not an accreditation as such, which could act as a strong trust anchor. From my perspective, from what we see so far, trust is working within the system (closed system) but there is no real trust anchor to the outside world other than the governance body being the provider and operator of the system. From my point of view there is an important link missing.

**Reply by mdoujak** [reply]
There is no difference with the PKI World. In PKI, the browser vendors decide which CAs they add to their trust lists. If the root certificate of the CA is in the Browser Trust List (or the Java Trust Store), then the TLS connection will work without interruptions. If the CA is missing, the Browser will start issuing warnings.
With SSI, the Wallet will either have the "root VC" of an ecosystem or not. The validation works if the root VC is present and the chain can be built. If the "root VC" is missing, the wallet will warn that the issuer/verifier cannot be validated.

The similarity can be taken one step further: With browsers, the browser vendor decides who belongs on the trust list. With the wallets, it is the issuer of the wallet that decides. If I am asked personally, I believe the Swiss government's vetting process is superior to the browser vendors' vetting process. I have some insights into the vetting processes of the browser vendors from roughly 20 years ago, and it was a mixture of commercial interests and usefulness to the end user community. I don't think this has changed very much.

Example: At least one Chinese Root CA is on the trust list. I am not sure why I should consider a Chinese CA server trustworthy. 

Example 2: With the advent of "Let's Encrypt", where ownership of a website is already sufficient to get a certificate, the trust implied by X.509 certificates has been reduced to the level of DNS.

Example 3: The "List of Trusted Lists" provided under eIDAS 1 needed to be processed to figure out, which CAs were allowed to issue certificates to PSD2 participants (TPP and PSP). A REST API was provided to download the member state provided lists. The lists themselves were not digitally signed and therefore authenticity of the lists could not be verified. At one point in time the REST API was moved to a new domain and the API was changed without notice. From what I understand the EU expects to continue with this trust model and every wallet will have to process the entire list to make the wallet interoperable with all issuers and verifiers in the EU. Considering the way this was managed in the past, I am not sure if I want to stake the security of the entire EUDI Wallet Ecosystems on this approach.

**Reply by AnAnalogGuy** [reply]
"There is no difference with the PKI World."
Technically I'm with you, which begs the question why not use this technology for the trust anchors (only) within the E-ID system. As with PKI, and that's my main point, there are governance, structure, organisation and processes that are not only well established and trusted, but also already operational, not just in the federal administration, but all over the world. Often you will find not only onboarding processes but also real accreditation business processes attached to a PKI that would provide trusted and auditable anchors for VCs. Why reinvent the wheel and build parallel and redundant processes and infrastructure?

**Reply by mdoujak** [reply]
There are several excellent reasons why using the PKI technology in SSI is a bad idea:
1. The PKI world has already proven that it cannot handle trust, which is precisely the point I was trying to make with my last reply.
2. The idea that a wallet (or any agent in SSI) must understand both VCs and X.509 certificates makes no sense. 
3. DIDs and VCs were invented to replace the monolithic world of X.509 certificates, and now we are supposed to continue to use them for the trust layer?!?
4. The X.509 world anchors trust with the TSP because the TSP has the CA certificate in the browser's trust store (Note: it takes years, almost decades, to propagate a new root CA to all the browsers in the world). With SSI I see this change. The trust anchor will be the owner of the ecosystem. I also expect the holder to be able to manage whether the wallet should accept a new ecosystem (i.e., a new trust anchor). Browsers have actively prevented this from being possible by overwriting every change in the trust list with every new version of the browser so that even the most diligent and security-aware person gave up trying to manage trust themselves.

BTW: I see no reason why ecosystem owners should not use TSPs with their processes and experience to handle the day-to-day operation of issuing issuer and verifier VCs.

**Reply by AnAnalogGuy** [reply]
I'm not talking about replacing VCs within the (closed) SSI system, but about using PKI to build the trust anchors to the outside world, mainly for the issuer and the verifier. Your request (to provide publicly readable and verifiable identity information) contradicts the basic concept of a VC in SSI, although it's absolutely necessary to build a trust anchor.

As the Swiss Federal Administration, to the best of my knowledge, is not planning to implement accreditation, but only onboarding of issuers, this link is missing, even if you would extend a VC with the relevant information. In this case the information is untrusted payload. I think the reason they don't plan a full accreditation service is because they know how complex and expensive it is, because it's 5% technology and 95% administration. And you automatically find yourself in a position of responsibility for your trustees, with all the legal and potentially financial consequences.

Yes, I'm with you, existing TSPs would be the best candidates to efficiently provide trust anchors. However, all their governance, processes, auditing, ... is strongly tied to the PKI. So it's not "just add some VCs". So again the question would be why not use "traditional" PKI for the trust anchor (which is already used in the VC signing process anyway) instead of introducing a (to some extent) proprietary solution. At least from a taxpayer's point of view, these scenarios should be investigated and calculated from a total cost of ownership point of view. 

Regarding the problems of adding X.509 certificates to the OS/browser's cert store: I'm aware of this, but it needs to be fixed where the problem originates. Building workarounds in dependent systems to fix problems in source systems has almost never been a good idea.

**Reply by mdoujak** [reply]
I think you did not get my point. I object to TSPs having the trust anchors. If we create an ecosystem then the trust anchor should be under the control of the governance body of the ecosystem and not some TSP. This would have the following advantages:

1. Many governance bodies already exist and they have leverage with their issuers (as opposed to a commercial TSP). This not only applies to issuing Issuer-VCs but also facilitates discussion about the semantic interoperability o the VCs that will be issued by the Issuers. E.g. FINMA has leverage with the banks and they can manage the process, how banks agree on the content of a VCs for "Bonität" or a VC for "Kontoverbindung" much more easily than a TSP.
2. The governance body will be liable for the issuers to some extent. This liability already exists today for existing governance bodies. I don't see how a TSP could take on this type of liability, which they must, if they want to be the trust anchor. E.g. if FINMA issued a banking license to a russian hacker, I am sure there would be some kind of liability for FINMA.

Secondly, the Swiss Federal Administration is already gearing up to issue VCs for the E-ID, driver's license, and so forth. Issuing a VC for an Issuer (or Verifier) is just another Schema, with some claims and a different approver. I don't see that this makes a huge difference on the technical side. 
Note: the number of issuers (and verifiers) is key. The level of automation required to handle the issuance will depend on the ecosystem. In finance I don't see a need for a high degree of automation since the total number of banks and FinTechs in Switzerland can still be managed manually. If we are talking about health care the number of ambulatory practices is in the tenthousands and automation would be needed. On the other hand FMH has with HIN already an organization that is perfectly placed to handle the issuance of issuer-VCs for doctors practices, should the need arise. 

Last thought: The first attempt to an E-ID was declined by the Swiss population because they did not want a commercial organization to manage all the E-IDs. I am not sure what the reaction would be, if we tried to tell them that now the same commercial organizations will manage the entire trust layer. 

I repeat: If a governance body decides to outsource the issuance of Issuer VCs to a TSP, I don't see a problem with that. But, the trust anchor remains with the governance body and the TSP just receives an Issuer-VC that allows the TSP to issue VCs to others.

**Reply by bjafs** [reply]
Hey there,

Thanks for your feedback!
Let me get back to you regarding the initial post.

At the moment, the trust protocol proposes a technical solution without delving too deeply into the governance of who operates it. The current legal framework assumes that the Confederation will manage a trust registry. Regarding additional third-party trust registries, there is no regulation within the e-ID bill.
Our current line of thought is that duplicating large amounts of information about entities from other existing registries is not highly beneficial and, moreover, complicates update mechanisms.
At its core, the value of the identity trust statement lies in the fact that the trust statement issuer has verified that the specific entity is who they claim to be. In our view, this is achieved by verifying that the person requesting the entry in the trust registry is authorized to act on behalf of the entity. Additionally, the entity needs to prove ownership of the Identifiers/Keys (DID-Document) hosted on the Confederation’s base registry.
By referencing the associated national registries (e.g., UID), our hypothesis was that additional information about the entity could be fetched from those sources. Information such as websites and legal references would most likely remain self-declared; therefore, we refrained from including them in the trust statement.

**Authority**
The current proposal is closely aligned with our regulatory requirements. It is relatively “flat” and does not currently foresee delegation to other authorizing bodies. This could be added at a later stage, or various sectorial “ecosystems” could exist in parallel. We are open to proposals on how this might evolve in the future.
In the current proposal, the issuer of the trust statement (in our case, the Swiss Confederation) serves as the authority verifying the information and issuing the statements. Since the authority signs the trust statements, its role is implicitly embedded in the trust VC. However, I believe I understand your point regarding chaining.

**Authorization**
If I understand you correctly, you are not referring to the proposed trust statements for issuance and verification, but rather suggesting that if VCs for authorities are introduced, information could be published regarding what they are allowed to attest. If we extend the concept, I believe it’s worth taking this into consideration.

**Schemas**
Schemas are mentioned in the Issuer and Verifier Trust Statements. Feel free to make a pull request or open an issue regarding the versioning. Currently, the hypothesis is that Issuer and Verifier Trust Statements and their associated schemas will primarily be offered for areas under the authority of the Confederation.
Regarding proof requests and presentation schemas, this is not a regulatory requirement at present. Evolution in this area could occur over time. However, since in Switzerland an entry in the trust registry is not mandatory, the system needs to function in a way that also works without this. To inform users about sensitive information in their credentials, a specific layer in OCAs will be used.

**Multiple Trust Registries in Wallets**
Yes, we see this as a possible evolution over time, but it is unlikely to be included at the initial launch of the system.

FYI: Most of us will be off during the festive season, so please bear with us if responses take a little longer than usual.

Best regards

**Reply by jurajsarinay** [reply]
>  a specific layer in OCAs will be used

Please avoid OCA. If unsure about this, find any computer scientist and show them the ["specification"](https://oca.colossi.network/specification/).
My favourite gems:

> - ***Numeric:*** a data type that defines anything relating to or containing numbers. Examples of numeric data types include 8-byte integers, fixed precision and scale numeric data, floating precision number data, integer (whole number) data, and monetary data values.
> - ***Boolean:*** a data type where the data only has two possible variables: true or false. In computer science, Boolean is an identification classifier for calculating logical truth values and algebraic variables.
>
> - ***Binary:*** a data type that defines a binary code signal, a series of electrical pulses representing numbers, characters, and performed operations. Based on a binary number system, each digit position represents a power of two (e.g., 4, 8, 16, etc.). A set of four binary digits or bits in binary code represents each decimal number (0 to 9). Each digit has two possible states: off and on (usually symbolised by 0 and 1). Combining basic Boolean algebraic operations on binary numbers makes it possible to represent each of the four fundamental arithmetic operations of addition, subtraction, multiplication, and division.

> There are many encoding standards including Base64 [[RFC4648](https://www.rfc-editor.org/info/rfc4648)], UTF-8, and ASCII to name a few. Each standard has a purpose, and applications using those encoding standards expect to receive data compliant with that encoding standard. The most popular types of character encoding are ASCII and Unicode [[UNICODE](https://home.unicode.org/)]. While ASCII is still supported by nearly all text editors, Unicode is more commonly used because it supports a larger character set. Unicode is often defined as UTF-8, UTF-16 [[RFC2781](https://www.rfc-editor.org/info/rfc2781)], or UTF-32 [[ISO10646](https://www.iso.org/standard/76835.html)], which refer to different Unicode standards.

![spec-fig1](https://oca.colossi.network/images/spec-fig1.png)
_Figure 1. In a balanced network, semantic overlays determine the meaning and use of what is put in, taken in, or operated on by any process or system._
![spec-fig2](https://oca.colossi.network/images/spec-fig2.png)
_Figure 2. In a balanced network, inputs overlays determine what is put in, taken in, or operated on by any process or system._

The FOITT should do better.

**Reply by bjafs** [reply]
Thank you for your message.

In my understanding, your input pertains to the handling of the visual representation rather than the trust protocol. This topic is already being discussed in thread #32, where you have also shared your concerns on OCA. It is requested that this thread remain specifically focused on the proposed trust protocol.

**Reply by jurajsarinay** [reply]
> your input pertains to the handling of the visual representation 

Not really. The real concern is a Swiss government agency happily implementing (or seriously considering) any random nonsense from the Internet. This would not be appropriate at BAG or BAZL (say). Yet for BIT, it seems to be business as usual. Anything goes.

**Comment by mdoujak** [comment]
A few thoughts...

**Trust Registry**
I understand that the Confederation will operate a trust registry, but
1. There is no statement about how many ecosystems == trust anchors can be added to this trust registry. I suggest building the base technology of the trust registry so that multiple trust anchors can be operated by different governance bodies. This would allow Fedpol to manage the E-ID ecosystem, and the cantonal MVDs could manage the driver's licenses.
2. The law allows the trust registry to be used for purposes other than the E-ID but does not make the federal trust registry mandatory. This implies that the law intends to permit the federal trust registry to coexist with other trust registries. If this was not supported by the law, ambition level 3 would not be achievable.

**Schemas**
Before we start the discussion on how to manage Schemas, I suggest that we agree on what we want to achieve with these Schemas. A few points would be:

1. Innovation - the solution must evolve to accommodate new or changed requirements.
2. Information—The solution must provide the holder with enough information to make an educated decision about the identity of issuers and verifiers.
3. quality - The solution must provide the holder with enough information to understand (1) how reliable data provided by an issuer is and (2) how reliable information about an issuer or a verifier is.
4. authorization - The solution must provide the holder with enough information to understand who registered/authorized this issuer.
5. over identification—The solution must inform the holder if a verifier requests more claims than agreed upon during registration.

I am sure that I missed quite a few requirements, but I wanted to add these thoughts to our discussion before I leave as well.

Enjoy the holiday season, and all the best for the new year!

mike

**Reply by bjafs** [reply]
Thanks Mike, 
We will get back to you on both points in the new year and will discuss taking the topic up for one of our next participation meetings. 
All the best and a happy festive season!

**Reply by bjafs** [reply]
**Hey mike,**

I hope you had a restful break!
As promised, I’m picking up on our discussion from last year.

**Regarding Trust Registries:**

- Yes, it is indeed possible for multiple trust registries to operate in parallel. 
- The e-ID Act neither regulates nor explicitly excludes third-party trust registries; it merely establishes a regulatory foundation for the trust registry operated by the Confederation.

The current technical proposal is relatively "flat" from a hierarchical perspective. 
However, it allows for various registries (e.g., sector- or domain-specific) to run in parallel. 
Within the trust registry managed by the Confederation, it is conceivable that authorization could be delegated to multiple authorities, allowing them to oversee governance surrounding their Verifiable Credentials (VCs). For example, ASA or Astra could act as the "owner" of the mDL, defining legitimate issuers (e.g., cantonal traffic authorities) and verifiers.

Additionally, within the Confederation’s trust registry, the confirmation of identities (e.g., ensuring that fedpol is who they claim to be) should largely be open to both private and public actors—provided certain stipulations are in place to ensure quality. This aligns with what is outlined in the e-ID Act.

The aforementioned claims of "legitimacy" will likely only be available for entities acting on behalf of the Confederation. This point should be clarified in the ordinance. 
Third-party registries could potentially address gaps in this area.

**Regarding Schemas:**
I believe we are broadly aligned on this point. It’s might be worth differentiating between information that describes the content of a credential and information that pertains to the use and governance of a credential.

The idea behind the proposed trust protocol is to use a specific VC issued by the trust registry to convey information about the legitimacy of issuers and verifiers, referencing a schema.

I find the concept of "quality" interesting. 
So far, my focus has primarily been on the actor level (verified vs. unknown). 
Do you have any ideas on how to address this at the content level?

Regarding point 5, Switzerland has so far taken a slightly different approach. 
Currently, there is no universal registration foreseen for verifiers. 
It is up to the "owner" of a VC to decide whether they want to issue verifier legitimacy credentials. 
In future, these credentials could be extended to include claims if needed.

The system is designed to support the "flagging" of sensitive information in a VC, warning users before transmitting this data to a verifier.
It will be up to the "owner" to define which of the data-fields are considered sensitive. 

I’d be happy to explore these points further in an in-person meeting, if that’s of interest to you.

Kind regards,
Andreas
